### ISSUE TITLE
Does not find Libraries and Matrices but they are there

### ISSUE THREAD

**Issue by stefandiederich** [GReD-Clermont/CLIFinder#1]
Hi, 

we try to run your CLIFinder script local on our server.

In the report.txt file it tells me:
```
alignement of iPSC_16-98_E to L1
alignement done
getting pairs with one mate matched to L1 and the other mate undetected by repeatmasker as a repeat sequence
The assumed RepeatMasker installation directory
    /BioinfSoftware/RepeatMasker
does not appear to be correct.  E.g it does not
contain a 'Libraries' or 'Matrices' subdirectory.
This can occur if hard links are used to invoke
this script.
cannot open /media/Berechnungen/CLIFinder-master//iPSC_16-98_E_2_repout/iPSC_16-98_E_2.fa.out No such file or directory
```

But there are subdirectories like `Libraries` and `Matrices` in the folder of RepeatMasker.

Can you please help me solving this issue?

Bests 
Stefan

**Comment by ppouchin** [comment]
@YRenaud Any idea what happens?

**Comment by ppouchin** [comment]
Which version of RepeatMasker do you have?
I'm not really familiar with CLIFinder, but I think the error message comes from RepeatMasker itself.

**Comment by stefandiederich** [comment]
Thanks for the quick response, I have RepeatMasker version 4.1.0

**Comment by ppouchin** [comment]
I'm sorry for this question, but do you have the same problem if you use version 4.0.3?

**Comment by stefandiederich** [comment]
Maybe I missed that, but I cant find a way to get version 4.0.3 on their homepage

**Comment by ppouchin** [comment]
Ok. It is indeed hard to find now. 
Have you tried installing the tool through the toolshed? 
https://toolshed.g2.bx.psu.edu/repository?repository_id=5c73d1cf20ab37c3

**Comment by ppouchin** [comment]
Ok. It seems the toolshed version is broken for now. I'll see if we can fix it. 
In the meantime, you can find older versions of RepeatMasker through conda, if you want to try the script with one of them (the earliest is 4.0.6).

**Comment by stefandiederich** [comment]
Hi, 
you mean the tooshed from Galaxy, right?
So we do not have a Galaxy instance installed at our local server. I tried to start the perl script directly from command line.

**Comment by ppouchin** [comment]
Ok. I'm trying to get some test data to check if I can get all the required tools.
If you have access to conda, could you try to create an environment like this?

```
conda create -n clifinder perl=5.22 perl-sys-c
pu perl-getopt-long perl-text-soundex perl-parallel-forkmanager perl-statistics-r r
-base r-plyr bioconductor-genomicranges bwa blast bedtools fastx_toolkit repeatmask
er
```

**Comment by ppouchin** [comment]
> 
> 
> Hi,
> 
> we try to run your CLIFinder script local on our server.
> 
> In the report.txt file it tells me:
> 
> ```
> alignement of iPSC_16-98_E to L1
> alignement done
> getting pairs with one mate matched to L1 and the other mate undetected by repeatmasker as a repeat sequence
> The assumed RepeatMasker installation directory
>     /BioinfSoftware/RepeatMasker
> does not appear to be correct.  E.g it does not
> contain a 'Libraries' or 'Matrices' subdirectory.
> This can occur if hard links are used to invoke
> this script.
> cannot open /media/Berechnungen/CLIFinder-master//iPSC_16-98_E_2_repout/iPSC_16-98_E_2.fa.out No such file or directory
> ```
> 
> But there are subdirectories like `Libraries` and `Matrices` in the folder of RepeatMasker.
> 
> Can you please help me solving this issue?
> 
> Bests
> Stefan

You have a "Libraries" folder, but have the RepeatMasker Libraries been installed there as specified here?
http://www.repeatmasker.org/RMDownload.html

**Comment by stefandiederich** [comment]
Hi,

sorry for my late response. We had some trouble with one of our server. So I had to fix that first... 

I will try to create the conda environment tomorrow. 
In my `Libraries` directory are the following files

```
drwxr-xr-x 2 die9s admins       4096 Dez 16 15:51 .
drwxr-xr-x 5 die9s admins       4096 Dez 16 15:51 ..
-rwxr-xr-- 1 die9s admins      25283 Dez 16 15:51 Artefacts.embl
-rwxr-xr-- 1 die9s admins   24005361 Dez 16 15:51 Dfam.embl
-rwxr-xr-- 1 die9s admins 1869701327 Dez 16 15:51 Dfam.hmm
-rwxr-xr-- 1 die9s admins        214 Dez 16 15:51 README.meta
-rwxr-xr-- 1 die9s admins   22475384 Dez 16 15:51 RepeatAnnotationData.pm
-rw-r--r-- 1 die9s admins   11365247 Dez 16 15:51 RepeatMasker.lib
-rw-r--r-- 1 die9s admins   22661790 Dez 16 15:51 RepeatMaskerLib.embl
-rwxr-xr-- 1 die9s admins   17979984 Dez 16 15:51 RepeatPeps.lib
-rwxr-xr-- 1 die9s admins       5550 Dez 16 15:51 RepeatPeps.readme
-rwxr-xr-- 1 die9s admins   18755326 Dez 16 15:51 RMRBMeta.embl
-rwxr-xr-- 1 die9s admins  113343436 Dez 16 15:51 taxonomy.dat
```

So I think they are installed properly. Or am I missing something?

Bests
Stefan

**Comment by ppouchin** [comment]
I finally got some data to try to run it. I was not involved in the development of this tool, so I don't really know how any of this works. However, I have access to the version that works on our Galaxy server for comparison. 
I'm currently trying to adapt the code so it could run from a conda environment (see the dev branch). 
For now, the conda environment I use is this setup like this: 
`conda create -n clifinder perl=5.22 perl-sys-cpu perl-getopt-long perl-text-soundex perl-parallel-forkmanager perl-statistics-r r-base r-plyr bioconductor-genomicranges bwa bedtools fastx_toolkit repeatmasker wget`
It should install repeatmasker 4.0.7 and I've changed the script so that the call to RepeatMasker uses the Dfam library. 
I got the tool to work without error, but I can't yet confirm that it produces good results as the small subset I used as input may not be suitable (and the whole data is about 30 GB).

**Comment by ppouchin** [comment]
So from what I've tested, if you use [the (currently) latest commit in dev](https://github.com/GReD-Clermont/CLIFinder/commit/71fdb23b7532f5ee170341494f3cdbd29fb876e5), and set up a conda environment matching the tool dependencies in the XML file, it should work.

The resulting conda environment should normally be created with:
`conda create -n clifinder samtools=1.9 bedtools=2.26.0gx repeatmasker=4.0.9_p2 bwa=0.7.17 fastx_toolkit=0.0.14 wget=1.20.1 perl=5.26 perl-getopt-long=2.50 perl-parallel-forkmanager=2.02 perl-statistics-r=0.34 r-base=3.5.1 r-plyr=1.8.5 bioconductor-genomicranges=1.34.0`

Please note that the tool has been adapted to use the `GALAXY_SLOTS` variable from Galaxy, so if you run it from the command line, you need to use the `--threads` option to run it on several CPU cores.

**Comment by stefandiederich** [comment]
Sorry for my late response and many many thanks for your help. I created a conda environment like you stated before. CLIFinder now does not throw any error message. But I get the following in the `report.txt` file:

```
alignement of iPSC_16-98E to L1
alignement done
getting pairs with one mate matched to L1 and the other mate undetected by repeatmasker as a repeat sequence
cannot open /media/Berechnungen/CLIFinder-dev//iPSC_16-98E_2_repout/iPSC_16-98E_2.fa.out No such file or directory
```

I do not think that this is an issue of CLIFinder itself. From the name of the directory I think it has to do with the repeatmasker output. the directory `iPSC_16-98E_2_repout` is there but there is no file in it. 
Maybe I gave some wrong paramter so repeatmasker does not find anything. Maybe there is something wrong with our TE file? 

Bests
Stefan

**Comment by ppouchin** [comment]
Maybe it doesn't find anything or maybe there's still something amiss. 
I've updated the program so it wouldn't rely on hardcoded files anymore. Although, you now have to provide the GenePred tables with the known repeat sequences and genes, as well as the blast databases (even though they are optional). 

You can still download the "refseq" file that CLIFinder used before from the URL in the code [here](https://github.com/GReD-Clermont/CLIFinder/blob/f6ec53e53ad89b5977f841424d2b7480a64133b6/LINE_html4.1.pl#L836), or you can use more up-to-date versions (for example if you use hg38). 
The rmsk file is easily obtainable from UCSC by downloading the "rmsk" track with all the columns. 

The argument for blast databases can be a path to a fasta file (with the `--build_[element]` argument), the path to a blast database (even in tar.gz form) or a URL to a tar.gz file (or even to a blast database with a wild-card parameter, e.g. https://www.site.com/rna*).

I've reworked the way the tool is used quite a bit so please tell me if something's not clear.

**Comment by stefandiederich** [comment]
Thanks for your "reimplementation" and the awsome help.

Can you please let me know where can I get the blast databases from?
I found and downloaded the other files already.

Bests
Stefan

**Comment by ppouchin** [comment]
You should be able to get the BLAST databases directly from the NCBI: RNA is [here](https://ftp.ncbi.nlm.nih.gov/blast/db/v4/refseq_rna_v4.01.tar.gz) (I think!) while EST should be [here ](https://ftp.ncbi.nlm.nih.gov/blast/db/v4/est_human_v4.00.tar.gz) and [here](https://ftp.ncbi.nlm.nih.gov/blast/db/v4/est_human_v4.01.tar.gz).
Otherwise, the script used to download them from our server as indicated [here](https://github.com/GReD-Clermont/CLIFinder/blob/dd4d7ef1985d311329899c3c20dbad3a9b3e0a9b/script/CLIFinder.pl#L77-L78).

**Comment by ppouchin** [comment]
If you go to the directory on our server containing the files CLIFinder used to download, I've created a subfolder named "[v0.5.0](https://galaxy.gred-clermont.fr/clifinder/v0.5.0/)" with all the corresponding files for this new version.

**Comment by stefandiederich** [comment]
Thanks for uploading the files! 

Somehow today I got the error message that `perl-file-copy-recursive` is missing. So I did a fresh clifinder conda environment with the addition of `perl-file-copy-recursive=0.45`. So maybe you can include that on the man page where you describe the way to build the conda environment. 

Right now the tool is running. I will give you feedback of any results :-)

Thanks a lot!
Bests Stefan

**Comment by stefandiederich** [comment]
Sorry, i did not want to close the issue, can you reopen it?

**Comment by ppouchin** [comment]
> 
> 
> Thanks for uploading the files!
> 
> Somehow today I got the error message that `perl-file-copy-recursive` is missing. So I did a fresh clifinder conda environment with the addition of `perl-file-copy-recursive=0.45`. So maybe you can include that on the man page where you describe the way to build the conda environment.
> 
> Right now the tool is running. I will give you feedback of any results :-)
> 
> Thanks a lot!
> Bests Stefan

Sorry about that. That's because I added the dependency since the last comment where I gave you the line to create the environment. 
If I'm not mistaken, it's already in the [README](https://github.com/GReD-Clermont/CLIFinder/blob/2a3a263fc5a2123e6988872d23057f8797231a76/README.rst), but I should have mentioned that you should recreate it.

**Comment by ppouchin** [comment]
FYI, I've made a small modification so that the temporary SAM file generated when aligning reads on TEs is (a lot) smaller. So if the process is taking too much disk space for you, you may want to try the latest commit from master.

**Comment by stefandiederich** [comment]
Sounds like a nice feature. Will try that in the next days.

CLIFinder runs good with your test data. Unfortunatelly when I run it with my data I got some error:

```
Alignment done
Getting pairs with one mate matched to L1 and the other mate undetected by repeatmasker as a repeat sequence
Number of half mapped pairs: 102602455
Could not run: /BioinfSoftware/Anaconda/envs/clifinder/bin/trf  /media/Berechnungen/CLIFinder-master/script/RM_60501.ThuFeb131801222020/iPSC_2.fa_batch-1.masked 2 7 7 80 10 50 10 : Cannot allocate memory
Cannot open /media/Berechnungen/CLIFinder-master/results//iPSC_2_repout/iPSC_2.fa.out No such file or directory
```

Sounds like there is not enough memory. But I am running on a server with 235GB RAM. Is it popssible, that CLIFinder uses all this?

Bests Stefan

**Comment by ppouchin** [comment]
Oh! Yes, it might be possible. That was one of the improvements I thought could be made: currently, all the half-mapped pairs are stored in memory.
I've never seen a problem on our server with 256 GB of RAM, but I've also not seen more than a few thousands half-mapped pairs...

How big are your input FASTQ files (in terms of order of magnitude)? Would you be able to try with a small subset?

The code might be changed, but I'm not at my office until February 24th...

**Comment by stefandiederich** [comment]
The fastq files contain arround 51,000,000 reads each. And I will give it a try with lets say a subset of  10,000 reads and will provide you with a feedback.

**Comment by stefandiederich** [comment]
So I runs properly now. I did some stupid mistake. Normaly we allways get *_R1.fastq (FWD) and *_R2.fastq (REV) files. But with this sequencing we had an extra file with unique indices. So the *_r3.fastq contains the reverse reads and R2 the 8 bp unique index. I started CLIFinder with R1 and R2 instead of R1 and R3. Therefore I had a lot of half-mapped pairs. 

So with the right files I now only have 5000 half-mapped reads. 
Thanks for all your help! Now we just have to interprete the results in the right way ;-)

**Comment by ppouchin** [comment]
Ok! Great!
So can I consider this issue closed?

**Comment by stefandiederich** [comment]
Yes you can :-)
Thanks a lot for all your help!

Bests Stefan

**Comment by ppouchin** [comment]
You're welcome! 
Good luck with your analysis!

**Comment by ppouchin** [comment]
@stefandiederich Sorry, it's not exactly about the initial problem, but I may have introduced a bug as I'm not an expert in Perl...
[Here](https://github.com/GReD-Clermont/CLIFinder/blob/master/script/CLIFinder.pl#L534-L536), I put "064" thinking it would be best for aligning lines, but Perl may consider this as an octal number and not a decimal... In which case it's not exactly doing what it should.
I'm also wondering if the line just before it is correct as aligned reads with a sam flag set to 101 or 165 are half-mapped but are not kept (although there may be a good reason to).
I'll put out a bugfix. ASAP.

**Comment by ppouchin** [comment]
> 
> 
> @stefandiederich Sorry, it's not exactly about the initial problem, but I may have introduced a bug as I'm not an expert in Perl...
> [Here](https://github.com/GReD-Clermont/CLIFinder/blob/master/script/CLIFinder.pl#L534-L536), I put "064" thinking it would be best for aligning lines, but Perl may consider this as an octal number and not a decimal... In which case it's not exactly doing what it should.
> I'm also wondering if the line just before it is correct as aligned reads with a sam flag set to 101 or 165 are half-mapped but are not kept (although there may be a good reason to).
> I'll put out a bugfix. ASAP.

Fixed by #7
